porch/pkg/git/package.go

```diff
@@ -152,36 +157,27 @@ func (p *gitPackageRevision) GetResources(ctx context.Context) (*v1alpha1.Packag
 }
 
 func (p *gitPackageRevision) GetUpstreamLock() (kptfile.Upstream, kptfile.UpstreamLock, error) {
-	repo, err := p.parent.getRepo()
+	resources, err := p.GetResources(context.Background())
 	if err != nil {
-		return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot determine package lock: %w", err)
+		return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot determine package lock; cannot retrieve resources: %w", err)
 	}
 
-	if p.ref == nil {
-		return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot determine package lock; package has no ref")
+	// get the upstream package URL from the Kptfile
+	var kf *kptfile.KptFile
+	for k, v := range resources.Spec.Resources {
+		if k == kptfile.KptFileName {
+			var err error
+			kf, err = pkg.DecodeKptfile(strings.NewReader(v))
+			if err != nil {
+				return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot read package lock from Kptfile: %w", err)
+			}
+		}
 	}
 
-	ref, err := refInRemoteFromRefInLocal(p.ref.Name())
-	if err != nil {
-		return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot determine package lock for %q: %v", p.ref, err)
+	if kf.Upstream == nil || kf.UpstreamLock == nil {
```

Comment:
Does this need to be an error? I don't think packages must have an upstream (some packages need to be on top of the hierarchy) and it seems reasonable that a package might not have the UpstreamLock either. If we think that the process of updating a package might be async, it would be possible to have a package with an upstream but no UpstreamLock.